packages/playwright/src/runner/loadUtils.ts

```diff
@@ -118,7 +121,50 @@ export async function loadFileSuites(testRun: TestRun, mode: 'out-of-process' |
   }
 }
 
-export async function createRootSuite(testRun: TestRun, errors: TestError[], shouldFilterOnly: boolean): Promise<Suite> {
+export async function detectChangedFiles(testRun: TestRun): Promise<string[]> {
+  const baseCommit = testRun.config.cliOnlyChanged;
+
+  function gitFileList(command: string) {
+    try {
+      return childProcess.execSync(
+          `git ${command}`,
+          { encoding: 'utf-8', stdio: 'pipe' }
+      ).split('\n').filter(Boolean);
+    } catch (_error) {
+      const error = _error as childProcess.SpawnSyncReturns<string>;
+      throw new Error([
+        `Encountered error while detecting changed files.`,
+        `--only-changed only works with Git repositories.`,
+        `Make sure that:`,
+        `  - You are running the test in a Git repository.`,
+        `  - The Git binary is in your PATH.`,
+        `  - The passed Git Ref exists in the repository. You passed '${baseCommit}'.`,
+        ``,
+        `Command Output:`,
+        ``,
+        ...error.output,
+      ].join('\n'));
+    }
+  }
+
+  const untrackedFiles = gitFileList(`ls-files --others --exclude-standard`).map(file => path.join(process.cwd(), file));
+
+  const [gitRoot] = gitFileList('rev-parse --show-toplevel');
+  const trackedFilesWithChanges = gitFileList(`diff ${baseCommit} --name-only`).map(file => path.join(gitRoot, file));
+
+  const filesWithChanges = [...untrackedFiles, ...trackedFilesWithChanges];
+
+  for (const plugin of testRun.config.plugins)
+    await plugin.instance?.populateDependencies?.();
```

Comment:
I'd move this call outside of this helper git-related function, probably to `tasks.ts`.